Add documentation on generation components (#308)
open-energy-transition/gb-dispatch-model: doc/gb-model/generators.rst

..
  SPDX-FileCopyrightText: Contributors to gb-dispatch-model <https://github.com/open-energy-transition/gb-dispatch-model>

  SPDX-License-Identifier: CC-BY-4.0

.. _generators:

##########################################
Generation Components
##########################################

This page describes how generation assets are represented in the model, including their data sources, capacity assignment, cost parameterisation, availability modelling, and implementation.

Overview
========

The generation system covers all electricity-producing assets connected to the modelled network, spanning both Great Britain and the surrounding European countries.
Generators are split into two broad groups:

- **Conventional generators**: Thermal and nuclear plant (CCGT, OCGT, gas engine, coal, nuclear, oil, waste, biomass, geothermal) that can be dispatched by the model within physical and contractual limits

> - **Renewable generators**: Solar, onshore wind, offshore wind, run-of-river hydro, and pumped hydro storage (PHS) whose output is constrained by weather-derived capacity-factor time series

Reviewer: Add a note that there is also generation using hydrogen turbines/fuel cells but they are handled in the hydrogen subsystem doc page (with a ref to that page)

Author: Thanks for the comment. I will provide details as suggested.

      enrich [label="Enriched powerplants\n(+ costs)", fillcolor="#FFFACD"];

      network [label="PyPSA Network\n(Generators / Stores)", fillcolor="#90EE90", shape=ellipse];

      fes -> ppl;
      dukes -> ppl;
      lcr -> ppl;
      ppl -> enrich;
      costs -> enrich;
      enrich -> network;
      entsoe -> avail -> network;
   }

.. _generators-data-sources:

Data Sources
============

FES BB1 — Future Capacity Projections
--------------------------------------

The primary source of *future* generator capacities for Great Britain is the NESO **Future Energy Scenarios (FES) 2024** workbook, sheet **BB1 (Building Blocks)**.

BB1 provides annual capacity projections (in GW) disaggregated to Grid Supply Points (GSPs) for each FES scenario.
Carrier and component-set assignment is performed through the ``fes.gb.carrier_mapping`` and ``fes.gb.set_mapping`` dictionaries in the configuration.

DUKES 5.11 — Existing Infrastructure
--------------------------------------

The **Digest of UK Energy Statistics (DUKES) table 5.11** provides a snapshot of *current* (2023) installed capacities for major power producers in Great Britain.
It is used to:

> 1. Anchor the spatial distribution of generators where FES does not provide GSP-level data

Reviewer: ```suggestion
1. Anchor the spatial distribution of generators where FES does not provide GSP-level data, only Transmission Operator (TO) level data.
```

We do still have some spatial disaggregation we work with, it's just much more coarse than GSP, so we use the DUKES data to distribute it further.

Author: Thank you for noting it.

+---------------------------+---------------------------------+
| ``OCGT``                  | Open-cycle gas turbine          |
+---------------------------+---------------------------------+
| ``engine``                | Gas reciprocating engine        |
+---------------------------+---------------------------------+
| ``nuclear``               | Nuclear (baseload)              |
+---------------------------+---------------------------------+
| ``coal``                  | Coal plant                      |
+---------------------------+---------------------------------+
| ``oil``                   | Oil-fired plant                 |
+---------------------------+---------------------------------+
| ``waste``                 | Waste incineration              |
+---------------------------+---------------------------------+
| ``biomass``               | Biomass/bioenergy               |
+---------------------------+---------------------------------+
| ``geothermal``            | Geothermal (minor capacity)     |
+---------------------------+---------------------------------+

Capacity is set to the FES BB1 projection for the model year, spatially distributed to model regions.

> A ``powerplants_filter`` (configured in ``electricity``) can exclude retired or out-of-scope plant.

Reviewer: Not sure we leverage this exclusion in our workflow. Can you check?

Author: Thank you for the comment. I was thinking about this too. We generally use functions from `add_electricity` to attach powerplants. However, filtering happens in `build_powerplants` which we do not utilize, as we have FES data. Thus, exclusion based on date is not in the model. I will remove such claim.

.. _generators-renewables:

Renewable Generators
---------------------

**PyPSA Component**: ``Generator`` with a time-varying ``p_max_pu`` profile

Renewable generators are capacity-constrained to an hourly capacity-factor time series derived from the atlite cutout and ERA5/SARAH reanalysis data via the standard PyPSA-Eur pipeline.

The renewable carriers are:

+---------------------------+------------------------------------------+
| Carrier                   | Notes                                    |
+===========================+==========================================+
| ``solar``                 | Utility-scale and rooftop PV             |
+---------------------------+------------------------------------------+
| ``onwind``                | Onshore wind                             |
+---------------------------+------------------------------------------+
| ``offwind-dc``            | Offshore wind (DC-connected)             |
+---------------------------+------------------------------------------+

> | ``hydro``                 | Run-of-river hydro                       |

Reviewer: ```suggestion
| ``hydro``                 | Reservoir hydro                       |
```

+---------------------------+------------------------------------------+

> | ``PHS``                   | Pumped hydro storage (``StorageUnit``)   |

Reviewer: Should this be in another, storage page?

Author: Thank you, I will put it into storage page.

+---------------------------+------------------------------------------+

.. _generators-chp:

> Combined Heat and Power (CHP)

Reviewer: Can you check that the most important features of the existing CHP docs are included here and then delete that section in the other page.

Reviewer: This brief implementation description is enough for our docs, compared to the far more verbose (and now likely outdated in some parts) section we added very early on

Author: Thank you, @brynpickering. I have reorganized CHP as suggested.

> Combined Heat and Power (CHP)

Reviewer: Mention how we define the capacity of CHP vs non-CHP generators (this comes from the FES)

Author: Hi, @brynpickering. Can you please clarify what do you mean by that?

Reviewer: There is a split between CHP and non-CHP versions of each technology (e.g. `CCGT PP` vs `CCGT CHP`). We get the relative capacity of `PP` vs `CHP` from the FES using the `set_mapping` key:

```yaml
    set_mapping:
      Technology:
        Biomass & Energy Crops (including CHP): CHP
        Waste Incineration (including CHP): CHP
        Non-renewable CHP: CHP
```

It's currently unclear in the text that only a subset of each technology is a CHP and how we differentiate in the workflow between them

- :math:`\delta_\text{shutdown}` — heat demand threshold below which the CHP may shut down completely (``chp.shutdown_threshold``, default 0.1)

This ensures CHPs run at least enough to cover heat demand, without modelling a separate heat bus.

.. _generators-availability:

Availability Modelling
======================

Monthly availability fractions are computed for GB thermal generators using ENTSO-E outage data.
The pipeline:

1. **Retrieves** planned and forced outage records from the ENTSO-E API by time period (chunked to stay within API rate limits)
2. **Filters** outages that exceed ``entsoe_unavailability.max_unavailable_days`` (default 366 days), treating very long outages as effectively retired plant rather than temporary unavailability
3. **Accumulates** unavailable MW on a 1-minute resolution time series for each carrier, then takes the monthly mean
4. **Calculates** availability fractions as:

.. math::

   a_m = \frac{C_\text{total} - \overline{U}_m}{C_\text{total}}

where :math:`C_\text{total}` is the total installed capacity for a carrier (from DUKES) and :math:`\overline{U}_m` is the mean unavailable capacity in month *m*.

> The resulting monthly profile is applied uniformly across all generators of the same carrier for the simulated year, scaling their effective ``p_nom``.

Reviewer: Add comment that this will be automatically applied even to European neighbours unless explicitly de-selected (at least, it will be available as an option once https://github.com/open-energy-transition/gb-dispatch-model/pull/307 is merged)

Author: Thank you. I will add details.

.. _generators-costs:

Cost Assignment
===============

Each generator in the powerplants table is enriched with cost and technical parameters through a multi-source lookup:

1. **FES AS.1**: Provides year-specific fuel costs and VOM for GB-relevant technologies; averaged across scenarios because scenario names change between FES editions
2. **FES AS.7**: Provides year-specific carbon price (£/tCO₂); averaged across scenarios

> 3. **PyPSA technology-data**: Provides capital costs (£/MW), fixed O&M (£/MW/year), efficiency, and lifetime for technologies absent from the FES costing workbook

Reviewer: Mention that the additional costs we get from the PyPSA technology are not actually used in the dispatch runs, they are just there for reference. We really only use efficiency from technology data in our current workflow.

Author: Thank you for noting it. I have added the changes.

The carriers included in the model and how they are treated (conventional vs renewable, generator vs store) are controlled by:

From ``config/config.gb.2024.yaml``:

.. literalinclude:: ../../config/config.gb.2024.yaml
   :language: yaml
   :start-after: # [doc:electricity-config-start]
   :end-before: # [doc:electricity-config-end]

CHP Constraints
----------------

From ``config/config.gb.2024.yaml``:

.. literalinclude:: ../../config/config.gb.2024.yaml
   :language: yaml
   :start-after: # [doc:chp-config-start]
   :end-before: # [doc:chp-config-end]

ENTSO-E Unavailability
-----------------------

From ``config/config.gb.2024.yaml``:

.. literalinclude:: ../../config/config.gb.2024.yaml
   :language: yaml
   :start-after: # [doc:entsoe-unavailability-config-start]
   :end-before: # [doc:entsoe-unavailability-config-end]

.. _generators-implementation-notes:

Implementation Notes
====================

**Data Processing Workflow**:

The generator system is built through a pipeline implemented in ``rules/gb-model/generators.smk``:

.. image:: img/generators_workflow.svg
   :align: center

.. note::
   The graph above was generated using::

      pixi run filtered_rulegraph \

>       "resources/GB/gb-model/HT/fes_powerplants_inc_tech_data.csv

Reviewer: Add all relevant files for generators that go into `rules/gb-model.smk::compose_network`. So, add the renewable profiles, tech costs file, hydro capacities(?). It might be worth documenting how we use pypsa-eur methods to actually prepare the network generator components, hence why we pass these pypsa-eur files to the final network composition step.

Author: Thanks for the comment. I have added details about other inputs of `compose_network` that are from PyPSA-Eur. However, adding them for plotting requires changing of `filter_gb_model_rulegraph.py`.

Reviewer: OK, good to know

1. GSP-level FES data for the same carrier and year
2. GSP-level DUKES data for the same carrier
3. GSP-level FES + DUKES data for *all* carriers combined
4. Overall national distribution as a fallback

This hierarchy preserves known regional concentrations (e.g., offshore wind clusters near specific GSPs) while gracefully handling carriers with no fine-grained spatial data.

**Key Assumptions**:

- **No capacity expansion**: All generator capacities are fixed at the FES scenario projections for the modelled year; the optimiser dispatches within these limits
- **Uniform availability**: Monthly availability fractions are applied uniformly to all generators of the same carrier across all regions
- **CHP heat coupling**: CHPs are constrained to run above a heat-demand-derived floor but are not connected to a separate heat bus; heat demand is satisfied exogenously
- **Cost averaging**: FES fuel and VOM costs are averaged across scenarios due to naming changes between FES editions; this has a small effect (≤ ~10% for battery VOM) on marginal costs

> - **European generators**: European countries use the PyPSA-Eur conventional generator database; availability and CHP constraints are not applied outside GB

Reviewer: availability _is_ applied outside GB unless switched off explicitly (new option in https://github.com/open-energy-transition/gb-dispatch-model/pull/307)

Author: Thank you for the comment.

> 

Reviewer: Add assumption that infinite fuel feedstocks are available

Author: Thank you for the suggestion. It is applied.

.. seealso::

   **Related Documentation**:

>    - :ref:`hydrogen-electrolysis` - Electrolysis as an additional flexible load

Reviewer: add a reference to hydrogen generators, not electrolysis

Author: Thank you, applied it.

>    - :doc:`ev` - EV demand and flexibility

Reviewer: no need

Author: Thank you, removed it.

CHPs are conventional generators (CCGT, OCGT, biomass, waste, etc.) that also supply heat.
Rather than implementing full sector-coupling, the model applies a simplified *minimum loading* constraint based on local heat demand.

> CHPs are identified from the powerplants table via the ``set`` column (``"CHP"``), which distinguishes them from regular power plants (``"PP"``) and storage units (``"Store"``).

Reviewer: ```suggestion
CHPs are identified from the FES based on generator names in sheet ``BB2`` and ``ES1`` mentioning ``CHP``.
We configure this identification under the ``set_mapping`` key in the configuration.
Identified CHPs then have the powerplants table ``set`` column value  ``"CHP"``, which distinguishes them from regular power plants (``"PP"``) and storage units (``"Store"``).
```

1. **FES AS.1**: Provides year-specific fuel costs and VOM for GB-relevant technologies; averaged across scenarios because scenario names change between FES editions
2. **FES AS.7**: Provides year-specific carbon price (£/tCO₂); averaged across scenarios

> 3. **PyPSA technology-data**: Provides capital costs (£/MW), fixed O&M (£/MW/year), efficiency, and lifetime for technologies absent from the FES costing workbook. In the current workflow only **efficiency** is actively used in dispatch runs; the remaining parameters are written to the output for reference only

Reviewer: ```suggestion
3. **PyPSA technology-data**: Provides capital costs (£/MW), fixed O&M (£/MW/year), efficiency, and lifetime for technologies absent from the FES costing workbook. 
   In the current workflow only **efficiency** is actively used in dispatch runs; the remaining parameters are written to the output for reference only
```
